### PR TITLE
blockdev: support installing to DM devices

### LINKED ISSUE #91: Non-partitionable target devices are not supported
# Bug #
I am unable to write an image to a block device, because coreos-installer fails to get an exclusive lock on the block device.

## Host Operating System Version ##
CentOS 7.7, kernel 3.10.0-1062.4.3.el7.x86_64

## Target Operating System Version ##
n/a

## coreos-installer Version ##
HEAD as of  Tue Nov 12 09:31:11 2019 -0500, commit hash: 48c4daf5ef50967a67e3f8ba78e9e788589e6130

## Expected Behavior ##
```
[16:14:08] sasami [/tmp] >> ./coreos-installer install /dev/mapper/storage-temp.as65342.net -s testing
Downloading testing image and signature
<verifying and writing a disk image>
```

## Actual Behavior ##
```
[16:14:08] sasami [/tmp] >> ./coreos-installer install /dev/mapper/storage-temp.as65342.net -s testing
Downloading testing image and signature
Error: checking for exclusive access to /dev/mapper/storage-temp.as65342.net
Caused by: rereading partition table
Caused by: EINVAL: Invalid argument
```

## Reproduction Steps ##
On centos 7:
```
yum install epel-release
yum install cargo openssl-devel
cargo build --release
./target/release/coreos-installer install /some/block/device -s testing
```

## Other Information ##
Trying to migrate my existing workflow based on CoreOS to FCOS, but running into this issue. This is with SELinux in enforcing mode, but this is not the cause for this bug.

### PR DESCRIPTION
**Pull request by bgilbert**
Handle partitioning with `kpartx`, and rely on the `lsblk` + `holders` busy-partition check to ensure exclusivity.

Fixes #91.  Alternative to #211.

### CODE REVIEW

**Comment by nikita-dubrovskii** [comment]
LGTM. 
Did test it on s390x

File: src/blockdev.rs [hunk 1]
```diff
@@ -132,6 +132,165 @@ impl Disk {
...
+impl PartTable for PartTableKpartx {
+    fn reread(&mut self) -> Result<()> {
+        self.run_kpartx("-u")
+    }
+}
+
+impl Drop for PartTableKpartx {
+    fn drop(&mut self) {
+        if self.need_teardown {
+            if let Err(e) = self.run_kpartx("-d") {
```
[the remark this answers is not in the document]
Comment (bgilbert) [reply src/blockdev.rs:289]: The idea is to leave the system in the same state we found it.  If we were given an idle non-partitionable device, and we run `kpartx -u` but not `kpartx -d`, we'll give back a device with DM devices attached to it.

Why an explicit method?  Letting the `PartTable` handle fall out of scope seems like a natural way to handle this, same as with a `File`.

**Comment by bgilbert** [comment]
Updated.

File: src/blockdev.rs [hunk 2]
```diff
@@ -132,6 +132,165 @@ impl Disk {
...
+        // Try rereading the partition table.  This is the most complete
+        // check, but it only works on partitionable devices.
+        let rereadpt_result = {
+            let mut f = OpenOptions::new()
+                .write(true)
+                .open(&self.path)
+                .chain_err(|| format!("opening {}", &self.path))?;
+            reread_partition_table(&mut f)
+        };
+        if rereadpt_result.is_ok() {
```
[the remark this answers is not in the document]
Comment (bgilbert) [reply src/blockdev.rs:138]: I tried:

```rust
        match rereadpt_result {
            Ok(_) => return rereadpt_result,
            Err(_) => (),
        }
```

but clippy said:
```
error: you seem to be trying to use match for destructuring a single pattern. Consider using `if let`
   --> src/blockdev.rs:138:9
    |
138 | /         match rereadpt_result {
139 | |             Ok(_) => return rereadpt_result,
140 | |             Err(_) => (),
141 | |         }
    | |_________^ help: try this: `if let Ok(_) = rereadpt_result { return rereadpt_result }`
    |
    = note: `-D clippy::single-match` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_match
```

I don't think that's cleaner than `is_ok()`, so I've switched back.

**Comment by jlebon** [comment]
Hadn't paid a lot of attention to this, but https://github.com/coreos/fedora-coreos-config/pull/503 is related here. I think it's possible to support this, but then it's the user's responsibility to make sure they add any necessary kargs (via `--append-karg`) for the machine to activate the right DM devices on boot. Because of this, I'm wondering if this should be gated behind some switch to make sure users understand this?

File: src/blockdev.rs [hunk 3]
```diff
@@ -173,7 +173,22 @@ impl Partition {
     }
 
     pub fn get_holders(&self) -> Result<Vec<String>> {
-        let holders = Path::new("/sys/block")
+        let holders = self.get_sysfs_dir()?.join("holders");
+        let mut ret: Vec<String> = Vec::new();
+        for ent in read_dir(&holders).chain_err(|| format!("reading {}", &holders.display()))? {
+            let ent = ent.chain_err(|| format!("reading {}", &holders.display()))?;
```
Comment (jlebon) [inline src/blockdev.rs:347]: Maybe something like

```suggestion
            let ent = ent.chain_err(|| format!("iterating {}", &holders.display()))?;
```

? So it's different from the `read_dir` error itself.

**Comment by bgilbert** [comment]
FWIW, #91 gives a case where FCOS can support this transparently: a DM device used as the boot disk for a VM.  The installer is run from the host, and the target device is non-partitionable from the host's perspective but partitionable from the guest's.

See also https://github.com/coreos/fedora-coreos-tracker/issues/321.

**Pull request merged**